### PR TITLE
Fixes  #522 : Basic test suite for OSGi framework

### LINKED ISSUE #522: Provide OSGi test project to validate OSGi headers automatically.
As stated in #82, we should provide a project for OSGi to validate correct exports.

### PR DESCRIPTION
**Pull request by theogimonde**
This is a proposal for a basic test project to ensure that the Mockito artifacts can be consumed from an OSGi environment. For example, reverting the fix for #678 is caught with the following exception:

```
java.lang.IllegalStateException: Failed to install bundle: mockito-core-3.2.3.jar
	at org.mockito.osgitest.OsgiTest.setUp(OsgiTest.java:55)
	... (snipped)
Caused by: org.osgi.framework.BundleException: Error occurred installing a bundle.
	at org.eclipse.osgi.storage.Storage.install(Storage.java:739)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.installBundle(BundleContextImpl.java:187)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.installBundle(BundleContextImpl.java:179)
	at org.mockito.osgitest.OsgiTest.setUp(OsgiTest.java:53)
	... 40 more
Caused by: java.lang.IllegalArgumentException: invalid range "[1.6.0": invalid format
	at org.osgi.framework.VersionRange.<init>(VersionRange.java:173)
	at org.eclipse.osgi.container.builders.OSGiManifestBuilderFactory.addPackageImports(OSGiManifestBuilderFactory.java:373)
	at org.eclipse.osgi.container.builders.OSGiManifestBuilderFactory.getPackageImports(OSGiManifestBuilderFactory.java:352)
	at org.eclipse.osgi.container.builders.OSGiManifestBuilderFactory.createBuilder(OSGiManifestBuilderFactory.java:111)
	at org.eclipse.osgi.container.builders.OSGiManifestBuilderFactory.createBuilder(OSGiManifestBuilderFactory.java:78)
	at org.eclipse.osgi.storage.Storage.getBuilder(Storage.java:784)
	at org.eclipse.osgi.storage.Storage.getBuilder(Storage.java:766)
	at org.eclipse.osgi.storage.Storage.install(Storage.java:706)
	... 43 more
Caused by: java.util.NoSuchElementException
	at java.util.StringTokenizer.nextToken(StringTokenizer.java:349)
	at org.osgi.framework.VersionRange.<init>(VersionRange.java:157)
	... 50 more
```

The OSGi test project consists of three source sets:

1. A regular JUnit4 `test` source set with a single class OsgiTest. Starts an OSGi framework, installs mockito-core, byte-buddy and objenesis bundles as well as two test bundles, and executes some simple Mockito test cases inside the OSGi framework.
2. A source set `testBundle` that's built to an OSGi bundle using the `Bundle` task from the Bnd Gradle Plugin. Contains the test classes that run the Mockito test cases and thus depends on the Mockito bundle.
3. A source set `otherBundle` that's also built to an OSGi bundle but with no other dependencies. Used for testing mocking of classes from other bundles.

Some further technical details:

- The `test` source set only has compile dependencies on the OSGi APIs and JUnit4, i.e. not on Mockito. The OSGi framework is created using the standard Framework API. At runtime Equinox is added as a dependency to supply an actual framework implementation, but the intention is that any framework implementation could be used.
- The test cases are run by loading classes from the test bundle that all implement `Runnable`. This is to provide a simple bridge using a common JDK class. The reason for setting up such a bridge is that I want to drive the tests using a test runner that Gradle supports, for example to get good test reports, but the test bundles don't know anything about JUnit4 inside the OSGi framework. I didn't want to explore adding an OSGified version of JUnit into the mix at this point since that would also entail solving test discovery inside an OSGi framework.
- The OSGi bundles are built using the [Bnd Gradle Plugin](https://github.com/bndtools/bnd/blob/master/biz.aQute.bnd.gradle/README.md#gradle-plugin-for-workspace-builds) instead of the deprecated `osgi` plugin in Gradle. The reason is both future proofing, to not impede a migration to Gradle 6.0 in this new sub project, and that I believe it better mirrors how other projects would consume the Mockito artifacts. Further it paves the way for replacing the usage of the deprecated plugin for the creating the OSGi manifests in the root project.

Since this is my first contribution I'll be happy to fix anything that goes against the grains.

### CODE REVIEW

File: subprojects/osgi-test/src/test/java/org/mockito/osgitest/OsgiTest.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+            }
+        } finally {
+            if (frameworkStorage != null) {
+                deleteRecursively(frameworkStorage);
+            }
+        }
+    }
+
+    @Test
+    public void testSimpleMock() throws BundleException, InterruptedException, ClassNotFoundException, IllegalAccessException, InstantiationException, IOException {
```
Comment (TimvdLippe) [inline subprojects/osgi-test/src/test/java/org/mockito/osgitest/OsgiTest.java]: Nit: let's make this `throws Exception` to be more concise.

File: subprojects/osgi-test/src/testBundle/java/org/mockito/osgitest/testbundle/MockClassInOtherBundleTest.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class MockClassInOtherBundleTest implements Runnable {
+
+    @Override
+    public void run() {
+        Methods methods = mock(Methods.class);
+        when(methods.intReturningMethod()).thenReturn(42);
+        if (methods.intReturningMethod() != 42) {
```
Comment (TimvdLippe) [inline subprojects/osgi-test/src/testBundle/java/org/mockito/osgitest/testbundle/MockClassInOtherBundleTest.java]: Nit: use `assertThat` from AssertJ or `assertThrows` from JUnit 5 to verify the exception. That makes the test clearer and provides better error messages when the test fails.

File: subprojects/osgi-test/src/testBundle/java/org/mockito/osgitest/testbundle/MockNonPublicClassFailsTest.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+public class MockNonPublicClassFailsTest implements Runnable {
+
+    static class NonPublicClass {}
+
+    @Override
+    public void run() {
+        try {
+            NonPublicClass nonPublicClass = mock(NonPublicClass.class);
+            throw new AssertionError();
+        } catch(MockitoException e) {
```
Comment (TimvdLippe) [inline subprojects/osgi-test/src/testBundle/java/org/mockito/osgitest/testbundle/MockNonPublicClassFailsTest.java]: Nit: don't use `try-catch` clauses in tests. Instead, use `assertThat` or `assertThrows`

File: subprojects/osgi-test/src/testBundle/java/org/mockito/osgitest/testbundle/MockClassInOtherBundleTest.java (hunk 2 above)
Comment (TimvdLippe) [reply subprojects/osgi-test/src/testBundle/java/org/mockito/osgitest/testbundle/MockClassInOtherBundleTest.java]: (In this case, an `assertEquals(methods.intReturningMethod(), 42)` would be better.

File: subprojects/osgi-test/src/test/java/org/mockito/osgitest/OsgiTest.java [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+        createTestRunnable("MockNonPublicClassFailsTest").run();
+    }
+
+    @Test
+    public void testMockClassInOtherBundle() throws BundleException, InterruptedException, ClassNotFoundException, IllegalAccessException, InstantiationException, IOException {
+        createTestRunnable("MockClassInOtherBundleTest").run();
+    }
+
+    private Runnable createTestRunnable(String className) throws InstantiationException, IllegalAccessException, ClassNotFoundException {
+        return (Runnable) testBundle.loadClass("org.mockito.osgitest.testbundle." + className).newInstance();
```
Comment (TimvdLippe) [inline subprojects/osgi-test/src/test/java/org/mockito/osgitest/OsgiTest.java]: Can we use the JUnit suite API to run these tests? For an example, see https://github.com/mockito/mockito/blob/release/3.x/src/test/java/org/mockitousage/junitrunner/StrictRunnerTest.java and the `JUnitCore` API.

**Comment by theogimonde** [comment]
> This is great work, thanks! Only some nits and a question whether we can use the JUnit API to handle the tests. That should improve the assertions and simplify the loading of the tests.

Thanks for the speedy review! It was intentional not to use the JUnit API since it's not readily available to consume as an OSGi bundle in this project. However, I have committed a new version where the `org.junit` package is provided to the OSGi bundles as a "system package" and thus is loaded from the regular class loader.

It's still tricky to get the test cases to run inside the bundles, since they aren't visible to Gradle in the `test` source set. In fact the test classes are only available during runtime _after_ starting the OSGi framework and installing their bundles. To overcome that issue I've inherited from `Suite` and fed it test classes loaded through the class loader of the test bundle.

Of course using proper asserts is preferable now that they are available so I fixed that. To get a clean diff from the first commit I did not yet consolidate the test classes, i.e. there's still one test class with a single `test()` method for each previous runnable, which might look a bit wasteful. Input on that is welcome, and also if you think that the added complexity of using a suite and the way JUnit is provided to the inside of OSGi is acceptable.

Technical details:
- The framework needs to be started in the suite constructor to be able to load the test classes. Thus `@BeforeClass` can no longer be used. However `@AfterClass` still works for the tear down.
- Using JUnit4 inside the test bundles makes Bnd add a corresponding `Import-Package: org.junit` to its manifest, the one that's resolved by the extra system package entry. This additional package import should not affect the effectiveness of the test for detecting manifest issues. In fact it might even make it more like other usages of Mockito.

File: subprojects/osgi-test/src/test/java/org/mockito/osgitest/OsgiTest.java (hunk 1 above)
Comment (theogimonde) [reply subprojects/osgi-test/src/test/java/org/mockito/osgitest/OsgiTest.java]: Makes sense. I collapsed multiple exception specifiers into `Exception` where applicable.

File: subprojects/osgi-test/src/testBundle/java/org/mockito/osgitest/testbundle/MockClassInOtherBundleTest.java (hunk 2 above)
Comment (theogimonde) [reply subprojects/osgi-test/src/testBundle/java/org/mockito/osgitest/testbundle/MockClassInOtherBundleTest.java]: Using `assertEquals(expected, actual)` now that it's available.

File: subprojects/osgi-test/src/test/java/org/mockito/osgitest/OsgiTest.java (hunk 4 above)
Comment (theogimonde) [reply subprojects/osgi-test/src/test/java/org/mockito/osgitest/OsgiTest.java]: I did use the suite API to make the tests run directly in the Gradle runner, which I think is better than using an explicit `JUnitCore` runner in this situation.

File: subprojects/osgi-test/src/testBundle/java/org/mockito/osgitest/testbundle/MockNonPublicClassFailsTest.java (hunk 3 above)
Comment (theogimonde) [reply subprojects/osgi-test/src/testBundle/java/org/mockito/osgitest/testbundle/MockNonPublicClassFailsTest.java]: Using `@Test(expected = MockitoException.class)` now that it's available.

**Pull request merged**